pkg/auth/client/impersonate.go

```diff
@@ -72,3 +62,51 @@ func NewImpersonatingKubernetesClientset(user user.Info, config restclient.Confi
 	impersonatingConfig := NewImpersonatingConfig(user, config)
 	return kclientset.NewForConfig(&impersonatingConfig)
 }
+
+// impersonatingRESTClient sets impersonating user, groups, and scopes headers per request
+type impersonatingRESTClient struct {
+	user     user.Info
+	delegate restclient.Interface
+}
+
+func NewImpersonatingRESTClient(user user.Info, client restclient.Interface) restclient.Interface {
+	return &impersonatingRESTClient{user: user, delegate: client}
+}
+
+// Verb does the impersonation per request by setting the proper headers
+func (c impersonatingRESTClient) Verb(verb string) *restclient.Request {
```

Comment:
Can this take a request as input so that methods like patch can become `return c.impersonate(c.delegate.Patch())`